### PR TITLE
ARROW-11156: [Rust][DataFusion] Create hashes vectorized in hash join

### PR DESCRIPTION
**Pull request by Dandandan**
Create hashes vectorized in hash join

This is one step for a fully vectorized hash join: https://issues.apache.org/jira/browse/ARROW-11112

The idea of the PR is as follows:

* We still use a `HashMap` but rather than using the row data as key we use a hash value ( `u64`) both as key and as hash. We use a custom `Hasher` to avoid (re)computing hashes in the hash map and while doing lookups.
* Only the hash value creation is in this PR vectorized, the rest is still on a row basis.
* A test for hash collision detection needs to be added.

TCPH 12 is without the remaining part ~10% faster than the other PR: ~180ms vs ~200ms.
TCPH 5 is >40% faster (332ms vs 624ms).

### CODE REVIEW

**Comment by jorgecarleitao** [comment]
👍  I think that this could be also beneficial for the hash aggregate!

**Comment by Dandandan** [comment]
The unit tests are succeeding now, had to do with the `RandomState` being reset on every `execute` 🎉

**Comment by Dandandan** [comment]
So what remains is the collision check.

**Comment by Dandandan** [comment]
@jorgecarleitao yes, definitely!

**Comment by Dandandan** [comment]
This is nearing readiness. Need to do some final cleanup & add a test for the collision detection.

Benchmarks results are still roughly the same.

**Comment by paddyhoran** [comment]
@Dandandan I was trying to set up the process for running the integration tests this morning and I cancelled your travis job by accident.  I'm not familiar with Travis, I looked to see if I could easily re-trigger it but I don't want to mess anything else up.

@jorgecarleitao @nevi-me @alamb @andygrove Is there an easy way to re-start just the travis job?

Sorry for the disruption.

**Comment by Dandandan** [comment]
@paddyhoran no problem, I will make some more changes anyway. Thanks for notifying!

File: rust/datafusion/src/physical_plan/hash_join.rs [hunk 1]
```diff
@@ -55,13 +56,12 @@ use crate::error::{DataFusionError, Result};
...
-use ahash::RandomState;
 use log::debug;
 
-// Maps ["on" value] -> [list of indices with this key's value]
-// E.g. [1, 2] -> [(0, 3), (1, 6), (0, 8)] indicates that (column1, column2) = [1, 2] is true
-// for rows 3 and 8 from batch 0 and row 6 from batch 1.
-type JoinHashMap = HashMap<Vec<u8>, Vec<u64>, RandomState>;
+// Maps a `u64` hash value based on the left ["on" values] to a list of indices with this key's value.
+// E.g. [1, 2] -> [3, 6, 8] indicates that the column values map to rows 3, 6 and 8
+// As the key is a hash value, we need to check possible hash collisions in the probe stage
```
Comment (alamb) [inline rust/datafusion/src/physical_plan/hash_join.rs:63]: 👍

**Comment by Dandandan** [comment]
This is ready for review now.

File: rust/datafusion/src/physical_plan/hash_aggregate.rs [hunk 2]
```diff
@@ -329,6 +329,76 @@ fn group_aggregate_batch(
     Ok(accumulators)
 }
 
+/// Create a key `Vec<u8>` that is used as key for the hashmap
+fn create_key(group_by_keys: &[ArrayRef], row: usize, vec: &mut Vec<u8>) -> Result<()> {
```
Comment (Dandandan) [inline rust/datafusion/src/physical_plan/hash_aggregate.rs:333]: This moved to `hash_aggregate`. It can move once hash aggregate does stops using `Vec<u8>` as a key

File: rust/datafusion/src/physical_plan/hash_join.rs [hunk 3]
```diff
@@ -80,6 +80,8 @@ pub struct HashJoinExec {
     schema: SchemaRef,
     /// Build-side
     build_side: Arc<Mutex<Option<JoinLeftData>>>,
+    /// Shares the `RandomState` for the hashing algorithm
+    random_state: RandomState,
```
Comment (Dandandan) [inline rust/datafusion/src/physical_plan/hash_join.rs:84]: This is needed to keep the random state the same between hashing and probing stage.

File: rust/datafusion/src/physical_plan/hash_join.rs [hunk 4]
```diff
@@ -193,6 +198,8 @@ impl ExecutionPlan for HashJoinExec {
     }
 
     async fn execute(&self, partition: usize) -> Result<SendableRecordBatchStream> {
+        let on_left = self.on.iter().map(|on| on.0.clone()).collect::<Vec<_>>();
```
Comment (Dandandan) [inline rust/datafusion/src/physical_plan/hash_join.rs:201]: Shared between build / probe stage.

File: rust/datafusion/src/physical_plan/hash_join.rs [hunk 5]
```diff
@@ -269,34 +276,35 @@ impl ExecutionPlan for HashJoinExec {
...
+            random_state: self.random_state.clone(),
         }))
     }
 }
 
 /// Updates `hash` with new entries from [RecordBatch] evaluated against the expressions `on`,
 /// assuming that the [RecordBatch] corresponds to the `index`th
 fn update_hash(
-    on: &HashSet<String>,
+    on: &[String],
```
Comment (Dandandan) [inline rust/datafusion/src/physical_plan/hash_join.rs:287]: `HashSet` is not needed

File: rust/datafusion/src/physical_plan/hash_join.rs [hunk 6]
```diff
@@ -575,6 +542,199 @@ fn build_join_indexes(
         }
     }
 }
+use core::hash::BuildHasher;
+
+/// `Hasher` that returns the same `u64` value as a hash, to avoid re-hashing
+/// it when inserting/indexing or regrowing the `HashMap`
+struct IdHasher {
+    hash: u64,
```
Comment (Dandandan) [inline rust/datafusion/src/physical_plan/hash_join.rs:550]: We just return the key value here, so we don't "rehash" when growing or indexing the `HashMap`

File: rust/datafusion/src/physical_plan/hash_join.rs [hunk 7]
```diff
@@ -575,6 +542,199 @@ fn build_join_indexes(
...
+impl BuildHasher for IdHashBuilder {
+    type Hasher = IdHasher;
+
+    fn build_hasher(&self) -> Self::Hasher {
+        IdHasher { hash: 0 }
+    }
+}
+
+// Combines two hashes into one hash
+fn combine_hashes(l: u64, r: u64) -> u64 {
```
Comment (Dandandan) [inline rust/datafusion/src/physical_plan/hash_join.rs:579]: Just a simple hashing combination function. Source: http://myeyesareblind.com/2017/02/06/Combine-hash-values/

Comment (Dandandan) [reply rust/datafusion/src/physical_plan/hash_join.rs:579]: Could be improved upon later.

File: rust/datafusion/src/physical_plan/hash_join.rs [hunk 8]
```diff
@@ -575,6 +542,199 @@ fn build_join_indexes(
...
+            == $r
+                .as_any()
+                .downcast_ref::<$array_type>()
+                .unwrap()
+                .value($right)
+    };
+}
+
+/// Left and right row have equal values
+fn equal_rows(
```
Comment (Dandandan) [inline rust/datafusion/src/physical_plan/hash_join.rs:599]: This is not vectorized (one row at a time). Could be vectorized in the future, checking an array of "matched candidates" instead of 1 by 1.

File: rust/datafusion/src/physical_plan/hash_join.rs [hunk 9]
```diff
@@ -907,4 +1068,53 @@ mod tests {
...
+            ("x", &vec![100, 200]),
+            ("y", &vec![200, 300]),
+        );
+
+        let random_state = RandomState::new();
+
+        let hashes = create_hashes(&[left.columns()[0].clone()], &random_state)?;
+
+        // Create hash collisions
+        hashmap_left.insert(hashes[0], vec![0, 1]);
```
Comment (Dandandan) [inline rust/datafusion/src/physical_plan/hash_join.rs:1086]: Maps to both indices, the test makes sure the extra "collisions" are removed.

File: rust/datafusion/src/physical_plan/hash_join.rs [hunk 10]
```diff
@@ -269,34 +276,35 @@ impl ExecutionPlan for HashJoinExec {
...
     // update the hash map
-    for row in 0..batch.num_rows() {
-        create_key(&keys_values, row, &mut key)?;
+    let hash_values = create_hashes(&keys_values, &random_state)?;
 
+    // insert hashes to key of the hashmap
+    for (row, hash_value) in hash_values.iter().enumerate() {
         hash.raw_entry_mut()
-            .from_key(&key)
+            .from_key_hashed_nocheck(*hash_value, hash_value)
```
Comment (Dandandan) [inline rust/datafusion/src/physical_plan/hash_join.rs:305]: hash = key now

**Comment by Dandandan** [comment]
FYI @jorgecarleitao @alamb @andygrove 
This PR is ready for review. I added some comments to the PR. My next step would be reusing the code/strategy for the hash aggregates. I think a similar speed up could be realized there as well.

**Comment by alamb** [comment]
Hi @Dandandan  thanks ! I will try and review this over the next few days.

File: rust/datafusion/src/physical_plan/hash_join.rs [hunk 11]
```diff
@@ -604,6 +562,199 @@ fn build_join_indexes(
...
+                )));
+                false
+            }
+        });
+
+    err.unwrap_or(Ok(res))
+}
+
+macro_rules! hash_array {
+    ($array_type:ident, $column: ident, $f: ident, $hashes: ident, $random_state: ident) => {
```
Comment (jorgecarleitao) [inline rust/datafusion/src/physical_plan/hash_join.rs:680]: IMO we should prob move this to the arrow crate at some point: hashing and equality are operations that need to be implemented with some invariants for consistency. I see that you had to implement an `equal_rows_elem`, which already indicates that.

File: rust/datafusion/src/physical_plan/hash_join.rs [hunk 12]
```diff
@@ -604,6 +562,199 @@ fn build_join_indexes(
...
+        let array = $column.as_any().downcast_ref::<$array_type>().unwrap();
+
+        for (i, hash) in $hashes.iter_mut().enumerate() {
+            let mut hasher = $random_state.build_hasher();
+            hasher.$f(array.value(i));
```
Comment (jorgecarleitao) [inline rust/datafusion/src/physical_plan/hash_join.rs:685]: Isn't this ignoring nulls? Maybe because they are discarded in joins? If so, isn't this a problem, as we are reading the value of a null slot?

**Review by jorgecarleitao — COMMENTED** [review]
I finally got the time to review this. IMO this looks great. Thanks a lot for taking this on!

I left some comments, all except one are optional / future PRs: I would really like your opinion wrt to the null values.

File: rust/datafusion/src/physical_plan/hash_join.rs [hunk 13]
```diff
@@ -604,6 +562,199 @@ fn build_join_indexes(
...
+            *hash = combine_hashes(hasher.finish(), *hash);
+        }
+    };
+}
+
+/// Creates hash values for every element in the row based on the values in the columns
+fn create_hashes(arrays: &[ArrayRef], random_state: &RandomState) -> Result<Vec<u64>> {
```
[the remark this answers is not in the document]
Comment (Dandandan) [reply rust/datafusion/src/physical_plan/hash_join.rs:700]: Not 100% sure how that would work for multiple columns? An earlier version of the PR also tried to reuse a bit of the allocation, but didn't seem to have a large impact on performance.
I think this can be looked at in a following PR. I am also wondering if there are some cool SIMD hash algorithms on multiple elements of arrays (rather than the common `Vec<u8>`) if we want to optimize`create_hashes`. Or maybe we can write one ourselves.  I am not sure how that would work with an iterator?

File: rust/datafusion/src/physical_plan/hash_join.rs (hunk 12 above)
Comment (Dandandan) [reply rust/datafusion/src/physical_plan/hash_join.rs]: I think so, I added some null handling in `hash_array` and `equal_rows_elem`. I think it makes sense that we have some more testing for nulls as well as this can be very tricky.
I am not sure this is handled 100% correctly with the previous code either, I think it is actually not correct, for example, the inner join doesn't check on null values. 

I added an issue here to add (more) tests on null values in keys:

https://issues.apache.org/jira/browse/ARROW-11258

File: rust/datafusion/src/physical_plan/hash_join.rs (hunk 11 above)
Comment (Dandandan) [reply rust/datafusion/src/physical_plan/hash_join.rs:679]: Yes, agreed.

File: rust/datafusion/src/physical_plan/hash_join.rs [hunk 14]
```diff
@@ -936,4 +1096,53 @@ mod tests {
...
+            ("a", &vec![10, 20]),
+            ("b", &vec![0, 0]),
+            ("c", &vec![30, 40]),
+        );
+
+        let left_data = JoinLeftData::new((hashmap_left, left));
+        let (l, r) = build_join_indexes(
+            &left_data,
+            &right,
+            JoinType::Inner,
```
Comment (alamb) [inline rust/datafusion/src/physical_plan/hash_join.rs:1127]: I recommend adding a test that covers collisions for all the `JoinType` variants -- the path for collision handling is different for those different join type variants.

**Review by alamb — APPROVED** [review]
Nice work @Dandandan  -- I reviewed the logic and the tests in this PR carefully (sorry it took so long) and they look good to me. I think the submodule issue (described below) needs to be sorted out before this is ready to merge. I think the test coverage for nulls could be improved (there is basically none now from what I can tell) but you have filed a follow on PR for that. 

It looks to me like there is a submodule update which may not be related to the changes contemplated in this PR -- I think we need to resolve:

Screen Shot 2021-01-15 at 1 56 58 PM

Otherwise this is looking great. I vote we ship it and continue improving joins / hashing later as follow on PRs.

**Comment by Dandandan** [comment]
Thanks @alamb submodule change reverted, not sure what happened there.

**Comment by alamb** [comment]
I apologize for the delay in merging Rust PRs -- the 3.0 release is being finalized now and are planning to minimize entropy by postponing merging  changes not critical for the release until the process was complete. I hope the process is complete in the next few days. There is more [discussion](https://lists.apache.org/list.html?dev@arrow.apache.org) in the mailing list

**Comment by alamb** [comment]
This one is next in line for merging @jorgecarleitao  and I have our eyes on it... Once a few more tests have completed on https://github.com/apache/arrow/commits/master we'll get it in

**Comment by jorgecarleitao** [comment]
I am calling the day, so feel free to continue the big flush. Thanks a lot for taking this, @alamb !

**Comment by alamb** [comment]
Master is looking pretty good: https://github.com/apache/arrow/runs/1729927062 and merged this branch locally into master and it compiles and passes tests. Merging it in
